File: C++/8-Graphs(BFS-DFS)/Algorithms/6-TopologicalSorting(Kahn’sAlgorithm)/TopologicalSorting(Kahn’sAlgorithm).js
/**
 * Topological Sorting (Kahn's Algorithm) - 5 Approaches in JavaScript
 *
 * Topological sorting for a Directed Acyclic Graph (DAG) is a linear ordering of vertices
 * such that for every directed edge u -> v, vertex u comes before vertex v in the ordering.
 * Kahn's algorithm is a popular algorithm for performing topological sorting.
 */

// Helper function to create an adjacency list from edge list
function createAdjacencyList(edges, numNodes) {
    const adjList = Array.from({ length: numNodes }, () => []);
    for (const [u, v] of edges) {
        adjList[u].push(v);
    }
    return adjList;
}

// Helper function to calculate in-degrees
function calculateInDegrees(adjList, numNodes) {
    const inDegrees = Array(numNodes).fill(0);
    for (let u = 0; u < numNodes; u++) {
        for (const v of adjList[u]) {
            inDegrees[v]++;
        }
    }
    return inDegrees;
}

// Example graph represented as an array of edges.  The nodes are assumed to be 0-indexed.
const edges = [
    [0, 1],
    [0, 2],
    [1, 3],
    [2, 3],
    [3, 4],
    [4, 5],
];
const numNodes = 6; // Number of nodes in the graph

// Approach 1: Basic Kahn's Algorithm using a queue
function topologicalSortKahnBasic(edges, numNodes) {
    const adjList = createAdjacencyList(edges, numNodes);
    const inDegrees = calculateInDegrees(adjList, numNodes);
    const queue = [];
    const result = [];

    // Enqueue nodes with in-degree 0
    for (let i = 0; i < numNodes; i++) {
        if (inDegrees[i] === 0) {
            queue.push(i);
        }
    }

    while (queue.length > 0) {
        const u = queue.shift();
        result.push(u);

        for (const v of adjList[u]) {
            inDegrees[v]--;
            if (inDegrees[v] === 0) {
                queue.push(v);
            }
        }
    }

    // Check for cycles (if the result doesn't contain all nodes, there's a cycle)
    if (result.length !== numNodes) {
        return "Graph contains a cycle. Topological sort is not possible.";
    }
    return result;
}

// Approach 2: Kahn's Algorithm with Priority Queue (for lexicographical order -  Smallest node first)
function topologicalSortKahnPriorityQueue(edges, numNodes) {
    const adjList = createAdjacencyList(edges, numNodes);
    const inDegrees = calculateInDegrees(adjList, numNodes);
    const pq = []; // Use a min-heap (priority queue)
    const result = [];

    // Enqueue nodes with in-degree 0
    for (let i = 0; i < numNodes; i++) {
        if (inDegrees[i] === 0) {
            pq.push(i);
        }
    }

    // Min-heap implementation (simple, for demonstration -  for larger graphs, use a proper heap library)
    pq.sort((a, b) => a - b); // Sort to maintain min-heap property

    while (pq.length > 0) {
        const u = pq.shift(); // Get the smallest element
        result.push(u);

        for (const v of adjList[u]) {
            inDegrees[v]--;
            if (inDegrees[v] === 0) {
                pq.push(v);
                pq.sort((a, b) => a - b); // Maintain min-heap property
            }
        }
    }

    if (result.length !== numNodes) {
        return "Graph contains a cycle. Topological sort is not possible.";
    }
    return result;
}

// Approach 3: Kahn's Algorithm with Set (for efficient removal)
function topologicalSortKahnSet(edges, numNodes) {
    const adjList = createAdjacencyList(edges, numNodes);
    const inDegrees = calculateInDegrees(adjList, numNodes);
    const zeroInDegreeSet = new Set();
    const result = [];

    // Add nodes with in-degree 0 to the set
    for (let i = 0; i < numNodes; i++) {
        if (inDegrees[i] === 0) {
            zeroInDegreeSet.add(i);
        }
    }

    while (zeroInDegreeSet.size > 0) {
        // Get an arbitrary element and remove it.  This is more efficient than
        // searching for the minimum in a long array.
        const u = zeroInDegreeSet.values().next().value;
        zeroInDegreeSet.delete(u);
        result.push(u);

        for (const v of adjList[u]) {
            inDegrees[v]--;
            if (inDegrees[v] === 0) {
                zeroInDegreeSet.add(v);
            }
        }
    }

    if (result.length !== numNodes) {
        return "Graph contains a cycle. Topological sort is not possible.";
    }
    return result;
}

// Approach 4: Kahn's Algorithm with Array as Queue (Optimized for speed)
function topologicalSortKahnArrayQueue(edges, numNodes) {
    const adjList = createAdjacencyList(edges, numNodes);
    const inDegrees = calculateInDegrees(adjList, numNodes);
    const queue = [];
    let queueIndex = 0; // Keep track of the "head" of the queue
    const result = [];

    // Enqueue nodes with in-degree 0
    for (let i = 0; i < numNodes; i++) {
        if (inDegrees[i] === 0) {
            queue.push(i);
        }
    }

    while (queueIndex < queue.length) { // Use queueIndex to track the "head"
        const u = queue[queueIndex++]; // Dequeue (efficiently)
        result.push(u);

        for (const v of adjList[u]) {
            inDegrees[v]--;
            if (inDegrees[v] === 0) {
                queue.push(v); // Enqueue
            }
        }
    }

    if (result.length !== numNodes) {
        return "Graph contains a cycle. Topological sort is not possible.";
    }
    return result;
}

// Approach 5: Kahn's Algorithm with Object as Queue (for sparse graphs) - less common, but shows variation
function topologicalSortKahnObjectQueue(edges, numNodes) {
    const adjList = createAdjacencyList(edges, numNodes);
    const inDegrees = calculateInDegrees(adjList, numNodes);
    const queue = {}; // Use an object as a queue (keys are indices, values are true)
    let head = 0;
    let tail = 0;
    const result = [];

    // Enqueue nodes with in-degree 0
    for (let i = 0; i < numNodes; i++) {
        if (inDegrees[i] === 0) {
            queue[tail++] = i; // Use tail to add to the "end"
        }
    }

    while (head < tail) { // While the "head" is behind the "tail"
        const u = queue[head++]; // Get from the "head" and move head forward
        result.push(u);

        for (const v of adjList[u]) {
            inDegrees[v]--;
            if (inDegrees[v] === 0) {
                queue[tail++] = v; // Add to the "end"
            }
        }
    }

    if (result.length !== numNodes) {
        return "Graph contains a cycle. Topological sort is not possible.";
    }
    return result;
}

// Display the results from each approach
console.log("Input Edges:", edges);
console.log("Number of Nodes:", numNodes);
console.log("\nTopological Sort Results:");
console.log("--------------------------");
console.log("Approach 1 (Basic Queue):", topologicalSortKahnBasic(edges, numNodes));
console.log("Approach 2 (Priority Queue):", topologicalSortKahnPriorityQueue(edges, numNodes));
console.log("Approach 3 (Set):", topologicalSortKahnSet(edges, numNodes));
console.log("Approach 4 (Array Queue):", topologicalSortKahnArrayQueue(edges, numNodes));
console.log("Approach 5 (Object Queue):", topologicalSortKahnObjectQueue(edges, numNodes));

File: Java/8-Graphs(BFS-DFS)/Algorithms/6-TopologicalSorting(Kahn’sAlgorithm)/TopologicalSorting(Kahn’sAlgorithm).java
import java.util.*;

public class TopologicalSort {

    // 1. Kahn's Algorithm using Array for In-degree and Queue
    //   - Uses a simple array to store in-degrees.
    //   - Iterates through the graph only once to calculate in-degrees.
    //   - Suitable for smaller graphs where the number of nodes is known and relatively small.
    public static List<Integer> topologicalSort1(int V, List<List<Integer>> adj) {
        int[] inDegree = new int[V]; // Array to store in-degrees of all vertices
        Queue<Integer> queue = new LinkedList<>(); // Queue to store vertices with in-degree 0
        List<Integer> result = new ArrayList<>(); // List to store the sorted vertices

        // Calculate in-degrees for each vertex
        for (int i = 0; i < V; i++) {
            for (Integer neighbor : adj.get(i)) {
                inDegree[neighbor]++;
            }
        }

        // Add vertices with in-degree 0 to the queue
        for (int i = 0; i < V; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }

        // Process vertices from the queue
        while (!queue.isEmpty()) {
            int u = queue.poll(); // Get the next vertex with in-degree 0
            result.add(u); // Add it to the result

            // Reduce in-degrees of its neighbors
            for (Integer neighbor : adj.get(u)) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.offer(neighbor); // If a neighbor's in-degree becomes 0, add it to the queue
                }
            }
        }

        // Check for cycles: If not all vertices are in the result, there's a cycle
        if (result.size() != V) {
            return new ArrayList<>(); // Return empty list to indicate cycle
        }

        return result;
    }

    // 2. Kahn's Algorithm using HashMap for In-degree and Queue
    //   - Uses a HashMap to store in-degrees, allowing for non-contiguous vertex labels.
    //   - Useful when vertex labels are not simple 0 to V-1 integers.
    //   - Slightly more memory overhead than using an array, but more flexible.
    public static List<Integer> topologicalSort2(int V, List<List<Integer>> adj) {
        Map<Integer, Integer> inDegree = new HashMap<>(); // HashMap to store in-degrees
        Queue<Integer> queue = new LinkedList<>(); // Queue to store vertices with in-degree 0
        List<Integer> result = new ArrayList<>(); // List to store the sorted vertices

        // Initialize in-degrees to 0 for all vertices
        for (int i = 0; i < V; i++) {
            inDegree.put(i, 0);
        }

        // Calculate in-degrees for each vertex
        for (int i = 0; i < V; i++) {
            for (Integer neighbor : adj.get(i)) {
                inDegree.put(neighbor, inDegree.get(neighbor) + 1); // Increment in-degree of neighbor
            }
        }

        // Add vertices with in-degree 0 to the queue
        for (Integer vertex : inDegree.keySet()) {
            if (inDegree.get(vertex) == 0) {
                queue.offer(vertex);
            }
        }

        // Process vertices from the queue
        while (!queue.isEmpty()) {
            int u = queue.poll(); // Get the next vertex with in-degree 0
            result.add(u); // Add it to the result

            // Reduce in-degrees of its neighbors
            for (Integer neighbor : adj.get(u)) {
                inDegree.put(neighbor, inDegree.get(neighbor) - 1); // Decrement in-degree of neighbor
                if (inDegree.get(neighbor) == 0) {
                    queue.offer(neighbor); // If a neighbor's in-degree becomes 0, add it to the queue
                }
            }
        }
        // Check for cycles: If not all vertices are in the result, there's a cycle
        if (result.size() != V) {
            return new ArrayList<>(); // Return empty list to indicate cycle
        }

        return result;
    }

    // 3. Kahn's Algorithm using Array and ArrayList for In-degree and Queue
    //   - Uses an array for in-degrees and an ArrayList as a queue.
    //   - Demonstrates using ArrayList as a queue (less efficient for queue operations, but possible).
    //   - Mainly for educational purposes to show different data structure options.
    public static List<Integer> topologicalSort3(int V, List<List<Integer>> adj) {
        int[] inDegree = new int[V]; // Array to store in-degrees
        List<Integer> queue = new ArrayList<>(); // ArrayList used as a queue
        List<Integer> result = new ArrayList<>(); // List to store the sorted vertices

        // Calculate in-degrees for each vertex
        for (int i = 0; i < V; i++) {
            for (Integer neighbor : adj.get(i)) {
                inDegree[neighbor]++;
            }
        }

        // Add vertices with in-degree 0 to the queue
        for (int i = 0; i < V; i++) {
            if (inDegree[i] == 0) {
                queue.add(i); // Add to the end of the list (enqueue)
            }
        }

        // Process vertices from the queue
        while (!queue.isEmpty()) {
            int u = queue.remove(0); // Remove from the front of the list (dequeue)
            result.add(u);

            for (Integer neighbor : adj.get(u)) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.add(neighbor); // Add to the end of the list (enqueue)
                }
            }
        }
        // Check for cycles: If not all vertices are in the result, there's a cycle
        if (result.size() != V) {
            return new ArrayList<>(); // Return empty list to indicate cycle
        }
        return result;
    }

    // 4. Kahn's Algorithm using PriorityQueue
    // - Uses PriorityQueue.
    // - The output order will be affected if there are multiple nodes with 0 in-degree.
    public static List<Integer> topologicalSort4(int V, List<List<Integer>> adj) {
        int[] inDegree = new int[V];
        PriorityQueue<Integer> pq = new PriorityQueue<>(); // Use PriorityQueue
        List<Integer> result = new ArrayList<>();

        // Calculate in-degrees for each vertex
        for (int i = 0; i < V; i++) {
            for (Integer neighbor : adj.get(i)) {
                inDegree[neighbor]++;
            }
        }

        // Add vertices with in-degree 0 to the queue
        for (int i = 0; i < V; i++) {
            if (inDegree[i] == 0) {
                pq.offer(i);
            }
        }

        // Process vertices from the queue
        while (!pq.isEmpty()) {
            int u = pq.poll();
            result.add(u);

            for (Integer neighbor : adj.get(u)) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    pq.offer(neighbor);
                }
            }
        }
        // Check for cycles: If not all vertices are in the result, there's a cycle
        if (result.size() != V) {
            return new ArrayList<>(); // Return empty list to indicate cycle
        }

        return result;
    }

    // 5. Kahn's Algorithm using LinkedList as Queue
    // - Uses LinkedList as the Queue.
    // - Demonstrates using LinkedList as queue.
    public static List<Integer> topologicalSort5(int V, List<List<Integer>> adj) {
        int[] inDegree = new int[V];
        LinkedList<Integer> queue = new LinkedList<>();
        List<Integer> result = new ArrayList<>();

        // Calculate in-degrees for each vertex
        for (int i = 0; i < V; i++) {
            for (Integer neighbor : adj.get(i)) {
                inDegree[neighbor]++;
            }
        }

        // Add vertices with in-degree 0 to the queue
        for (int i = 0; i < V; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i); //offer is used to add element in queue
            }
        }

        // Process vertices from the queue
        while (!queue.isEmpty()) {
            int u = queue.poll();  //poll() is used to remove the head of the queue
            result.add(u);

            for (Integer neighbor : adj.get(u)) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        // Check for cycles: If not all vertices are in the result, there's a cycle
        if (result.size() != V) {
            return new ArrayList<>(); // Return empty list to indicate cycle
        }

        return result;
    }

    public static void main(String[] args) {
        int V = 6; // Number of vertices
        // Adjacency list representation of the graph
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }

        // Add edges to the graph (0-based indexing)
        adj.get(5).add(2);
        adj.get(5).add(0);
        adj.get(4).add(0);
        adj.get(4).add(1);
        adj.get(2).add(3);
        adj.get(3).add(1);

        // Perform topological sort using different methods and print the results
        System.out.println("Topological Sort (Method 1 - Array, Queue): " + topologicalSort1(V, adj));
        System.out.println("Topological Sort (Method 2 - HashMap, Queue): " + topologicalSort2(V, adj));
        System.out.println("Topological Sort (Method 3 - Array, ArrayList): " + topologicalSort3(V, adj));
        System.out.println("Topological Sort (Method 4 - Array, PriorityQueue): " + topologicalSort4(V, adj));
        System.out.println("Topological Sort (Method 5 - Array, LinkedList): " + topologicalSort5(V, adj));

        // Example of a graph with a cycle (for testing cycle detection)
        List<List<Integer>> cyclicAdj = new ArrayList<>();
        for (int i = 0; i < 3; i++) {
            cyclicAdj.add(new ArrayList<>());
        }
        cyclicAdj.get(0).add(1);
        cyclicAdj.get(1).add(2);
        cyclicAdj.get(2).add(0); // This creates a cycle: 0 -> 1 -> 2 -> 0

        System.out.println("Topological Sort of Cyclic Graph (Method 1): " + topologicalSort1(3, cyclicAdj)); // Should return []
    }
}

